core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/ldap/LDAPConnectionContext.java

```diff
@@ -203,10 +209,87 @@ public LDAPConnectionContext(RealmConfiguration realmConfig) throws UserStoreExc
         // Set StartTLS option if provided in the configuration. Otherwise normal connection.
         startTLSEnabled = Boolean.parseBoolean(realmConfig.getUserStoreProperty(
                 UserStoreConfigConstants.STARTTLS_ENABLED));
+
+        // Set waiting time to re-establish LDAP connection after couple of failure attempts if specified otherwise
+        // set default time.
+        String retryWaitingTime = realmConfig.getUserStoreProperty(UserStoreConfigConstants.CONNECTION_RETRY_DELAY);
+        if (StringUtils.isNotEmpty(retryWaitingTime)) {
+            thresholdTimeoutInMilliseconds = getThresholdTimeoutInMilliseconds(retryWaitingTime);
+        } else {
+            thresholdTimeoutInMilliseconds = 120000;
+        }
+        // By-default set to close state.
+        ldapConnectionCircuitBreakerState = CIRCUIT_STATE_CLOSE;
+        thresholdStartTime = 0;
     }
 
     public DirContext getContext() throws UserStoreException {
 
+        DirContext context = null;
+
+        // Implemented basic circuit breaker logic to reduce the resource consumption.
+        switch (ldapConnectionCircuitBreakerState) {
+        case CIRCUIT_STATE_OPEN:
+            long circuitOpenDuration = System.currentTimeMillis() - thresholdStartTime;
+            if (circuitOpenDuration >= thresholdTimeoutInMilliseconds) {
+                try {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Trying to obtain LDAP connection, connection URL: " + environment
```

Comment:
Can the 'environment'' variable be null? if so possible NPE in the debug logs

Reply:
the "environment" is got initialised with this class initialization constructor, so it won't be null.